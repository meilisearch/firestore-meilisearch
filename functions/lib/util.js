'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChangedDocumentId = exports.getActionType = exports.ChangeType = void 0;
var ChangeType;
(function (ChangeType) {
    ChangeType[ChangeType["CREATE"] = 0] = "CREATE";
    ChangeType[ChangeType["DELETE"] = 1] = "DELETE";
    ChangeType[ChangeType["UPDATE"] = 2] = "UPDATE";
})(ChangeType = exports.ChangeType || (exports.ChangeType = {}));
/**
 * Get type of the modification perform on a document.
 * @param {Change<T>} change The Functions interface for events that change state.
 * @return {ChangeType} Final type of the event.
 */
function getActionType(change) {
    if (!change.after.exists) {
        return ChangeType.DELETE;
    }
    if (!change.before.exists) {
        return ChangeType.CREATE;
    }
    return ChangeType.UPDATE;
}
exports.getActionType = getActionType;
/**
 * Get final id of a document after modification.
 * @param {Change<T>} change The Functions interface for events that change state.
 * @return {string} Final state type of the event.
 */
function getChangedDocumentId(change) {
    if (change.after.exists) {
        return change.after.id;
    }
    return change.before.id;
}
exports.getChangedDocumentId = getChangedDocumentId;
