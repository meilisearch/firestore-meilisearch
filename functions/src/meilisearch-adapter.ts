'use strict'
/*
 * Copyright 2022 Meilisearch
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DocumentSnapshot } from 'firebase-functions/lib/v1/providers/firestore'
import * as firestore from 'firebase-admin/firestore'
import { infoGeoPoint } from './logs'

type MeilisearchGeoPoint = {
  lat: number
  lng: number
}

/**
 * Adapts GeoPoint Firestore instance to fit with Meilisearch geo point.
 * @param {firestore.GeoPoint} geoPoint GeoPoint Firestore object.
 * @return {MeilisearchGeoPoint} A properly formatted geo point for Meilisearch.
 */
function adaptGeoPoint(geoPoint: firestore.GeoPoint): MeilisearchGeoPoint {
  return {
    lat: geoPoint.latitude,
    lng: geoPoint.longitude,
  }
}

/**
 * Check if the field is added to the document send to Meilisearch.
 *
 * @param  {string[]} fieldsToIndex
 * @param  {string} key
 * @return {boolean} true if it is a field that should be indexed in Meilisearch
 *
 */
export function isAFieldToIndex(fieldsToIndex: string[], key: string): boolean {
  if (
    fieldsToIndex.length === 0 ||
    fieldsToIndex.includes('*') ||
    fieldsToIndex.includes(key)
  ) {
    return true
  }
  return false
}

/**
 * Parse the fieldsToIndex string into an array.
 *
 * @param  {string} fieldsToIndex
 * @return {string[]} An array of fields.
 */
export function parseFieldsToIndex(fieldsToIndex: string): string[] {
  return fieldsToIndex ? fieldsToIndex.split(/[ ,]+/) : []
}

/**
 * Update special fields to Meilisearch compatible format
 * @param {firestore.DocumentData} document
 * @param {string[]} fieldsToIndexSetting
 * @return {firestore.DocumentData} A properly formatted array of field and value.
 */
export function adaptFields(
  document: firestore.DocumentData,
  fieldsToIndexSetting: string
): firestore.DocumentData {
  const fieldsToIndex = parseFieldsToIndex(fieldsToIndexSetting)

  return Object.keys(document).reduce((doc, currentField) => {
    const value = document[currentField]

    if (!isAFieldToIndex(fieldsToIndex, currentField)) return doc
    if (value instanceof firestore.GeoPoint) {
      if (currentField === '_geo') {
        infoGeoPoint(true)
        return {
          ...doc,
          _geo: adaptGeoPoint(value),
        }
      } else {
        infoGeoPoint(false)
      }
    }
    return { ...doc, [currentField]: value }
  }, {})
}

/**
 * Adapts documents from the Firestore database to Meilisearch compatible documents.
 * @param {string} documentId Document id.
 * @param {DocumentSnapshot} snapshot Snapshot of the data contained in the document read from your Firestore database.
 * @param {string} fieldsToIndexSetting Value of the setting `FIELDS_TO_INDEX`
 * @return {Record<string, any>} A properly formatted document to be added or updated in Meilisearch.
 */
export function adaptDocumentForMeilisearch(
  documentId: string,
  snapshot: DocumentSnapshot,
  fieldsToIndexSetting: string
): Record<string, any> {
  const data = snapshot.data() || {}
  if ('_firestore_id' in data) {
    delete data.id
  }
  const adaptedDoc = adaptFields(data, fieldsToIndexSetting)

  return { _firestore_id: documentId, ...adaptedDoc }
}
