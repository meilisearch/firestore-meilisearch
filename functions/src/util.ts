'use strict'
/*
 * Copyright 2022 Meilisearch
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DocumentSnapshot } from 'firebase-functions/lib/providers/firestore'
import { Change } from 'firebase-functions'

export enum ChangeType {
  CREATE,
  DELETE,
  UPDATE,
}

/**
 * Get type of the modification perform on a document.
 * @param {Change<T>} change The Functions interface for events that change state.
 * @return {ChangeType} Final type of the event.
 */
export function getActionType(change: Change<DocumentSnapshot>): ChangeType {
  if (!change.after.exists) {
    return ChangeType.DELETE
  }
  if (!change.before.exists) {
    return ChangeType.CREATE
  }
  return ChangeType.UPDATE
}

/**
 * Get final id of a document after modification.
 * @param {Change<T>} change The Functions interface for events that change state.
 * @return {string} Final state type of the event.
 */
export function getChangedDocumentId(change: Change<DocumentSnapshot>): string {
  if (change.after.exists) {
    return change.after.id
  }
  return change.before.id
}
