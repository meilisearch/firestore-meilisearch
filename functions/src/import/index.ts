#!/usr/bin/env node

/*
 * Copyright 2022 Meilisearch
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as admin from 'firebase-admin'
import { DocumentSnapshot } from 'firebase-functions/lib/providers/firestore'
import { CLIConfig, parseConfig } from './config'
import * as logs from '../logs'
import { adaptDocumentForMeilisearch } from '../meilisearch-adapter'
import { initMeilisearchIndex } from '../meilisearch/create-index'
import { Index } from '../types'

const run = async () => {
  // Retrieve all arguments from the commande line.
  const config: CLIConfig = await parseConfig()

  // Initialize Firebase using the Google Credentials in the GOOGLE_APPLICATION_CREDENTIALS environment variable.
  admin.initializeApp({
    credential: admin.credential.applicationDefault(),
    databaseURL: `https://${config.projectId}.firebaseio.com`,
  })

  const database = admin.firestore()

  // Initialize Meilisearch index.
  const index = initMeilisearchIndex(config.meilisearch)

  await retrieveCollectionFromFirestore(database, config, index)
}

/**
 * Retrieves a collection or collection group in Firestore and aggregates the data.
 * @param {FirebaseFirestore.Firestore} database
 * @param {CLIConfig} config
 * @param {Index} index
 */
async function retrieveCollectionFromFirestore(
  database: FirebaseFirestore.Firestore,
  config: CLIConfig,
  index: Index
): Promise<number> {
  const batch: number = parseInt(config.batchSize)

  let query
  let total = 0
  let batches = 0
  let lastDocument = null
  let lastBatchSize: number = batch

  while (lastBatchSize === batch) {
    batches++

    if (config.queryCollectionGroup) {
      query = database.collectionGroup(config.sourceCollectionPath).limit(batch)
    } else {
      query = database.collection(config.sourceCollectionPath).limit(batch)
    }

    if (lastDocument !== null) {
      query = query.startAfter(lastDocument)
    }

    const snapshot = await query.get()
    const docs = snapshot.docs

    if (docs.length === 0) break
    total += await sendDocumentsToMeilisearch(
      docs,
      index,
      config.meilisearch?.fieldsToIndex || ''
    )

    if (docs.length) {
      lastDocument = docs[docs.length - 1]
    }
    lastBatchSize = docs.length
  }

  logs.importData(total, batches)
  return total
}

/**
 * Adapts documents and indexes them in Meilisearch.
 * @param {Change<DocumentSnapshot>} docs
 * @param {Index} index
 * @param {string} fieldsToIndex list of fields added in the document send to Meilisearch.
 */
async function sendDocumentsToMeilisearch(
  docs: DocumentSnapshot[],
  index: Index,
  fieldsToIndex: string
): Promise<number> {
  const document = docs.map(snapshot => {
    return adaptDocumentForMeilisearch(snapshot.id, snapshot, fieldsToIndex)
  })
  try {
    await index.addDocuments(document, { primaryKey: '_firestore_id' })
  } catch (e) {
    logs.error(e as Error)
    return 0
  }
  return document.length
}

void run()
