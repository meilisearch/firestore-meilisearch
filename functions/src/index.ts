'use strict'
/*
 * Copyright 2022 Meilisearch
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as functions from 'firebase-functions'
import { Change, logger } from 'firebase-functions'
import { DocumentSnapshot } from 'firebase-functions/lib/providers/firestore'
import { initMeilisearchIndex } from './meilisearch/create-index'
import { getActionType, getChangedDocumentId, ChangeType } from './util'
import * as logs from './logs'
import { adaptDocumentForMeilisearch } from './meilisearch-adapter'
import { config } from './config'
import { validateDocumentId } from './validate'

const index = initMeilisearchIndex(config.meilisearch)

logs.init()

/**
 * IndexingWorker is responsible for aggregating a defined field from a Firestore collection into a Meilisearch index.
 * It is controlled by a Firestore handler.
 */
export const indexingWorker = functions.handler.firestore.document.onWrite(
  async (snapshot: Change<DocumentSnapshot>): Promise<void> => {
    logs.start()
    const actionType = getActionType(snapshot)
    const documentId = getChangedDocumentId(snapshot)

    switch (actionType) {
      case ChangeType.CREATE:
        await handleAddDocument(documentId, snapshot.after)
        break
      case ChangeType.DELETE:
        await handleDeleteDocument(documentId)
        break
      case ChangeType.UPDATE:
        await handleUpdateDocument(documentId, snapshot.after)
        break
    }
    logs.complete()
  }
)

/**
 * Handle addition of a document in the Meilisearch index.
 * @param {string} documentId Document id to add.
 * @param {Change} snapshot Snapshot of the data contained in the document read from your Firestore database.
 */
async function handleAddDocument(
  documentId: string,
  snapshot: DocumentSnapshot
): Promise<void> {
  try {
    logs.addDocument(documentId)
    if (validateDocumentId(documentId)) {
      const document = adaptDocumentForMeilisearch(
        documentId,
        snapshot,
        config.meilisearch.fieldsToIndex || ''
      )
      const { taskUid } = await index.addDocuments([document], {
        primaryKey: '_firestore_id',
      })

      logger.info(
        `Document addition request for document with ID ${documentId} added to task list (task ID ${taskUid}).`
      )
    } else {
      logger.error(
        `Could not create document with id: ${documentId}. The document id can only contain case-insensitive alphanumeric characters (abcDEF), hyphens (-) or underscores(_).`
      )
    }
  } catch (e) {
    logs.error(e as Error)
  }
}

/**
 * Handle deletion of a document in the Meilisearch index.
 * @param {string} documentId Document id to delete.
 */
async function handleDeleteDocument(documentId: string): Promise<void> {
  try {
    logs.deleteDocument(documentId)
    if (validateDocumentId(documentId)) {
      const { taskUid } = await index.deleteDocument(documentId)

      logger.info(
        `Document deletion request for document with ID ${documentId} added to task list (task ID ${taskUid}).`
      )
    } else {
      logger.error(
        `Could not delete document with id: ${documentId}. The document id can only contain case-insensitive alphanumeric characters (abcDEF), hyphens (-) or underscores(_).`
      )
    }
  } catch (e) {
    logs.error(e as Error)
  }
}

/**
 * Handle update of a document in the Meilisearch index.
 * @param {string} documentId Document id to update.
 * @param {Change} after Snapshot of the data contained in the document read from your Firestore database.
 */
async function handleUpdateDocument(
  documentId: string,
  after: DocumentSnapshot
): Promise<void> {
  try {
    logs.updateDocument(documentId)
    if (validateDocumentId(documentId)) {
      const document = adaptDocumentForMeilisearch(
        documentId,
        after,
        config.meilisearch.fieldsToIndex || ''
      )
      const { taskUid } = await index.addDocuments([document])

      logger.info(
        `Document update request for document with ID ${documentId} added to task list (task ID ${taskUid}).`
      )
    } else {
      logger.error(
        `Could not update document with id: ${documentId}.The document id can only contain case-insensitive alphanumeric characters (abcDEF), hyphens (-) or underscores(_).`
      )
    }
  } catch (e) {
    logs.error(e as Error)
  }
}
