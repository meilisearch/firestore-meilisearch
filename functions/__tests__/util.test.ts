import * as firebaseFunctionsTestInit from 'firebase-functions-test'
import mockedEnv from 'mocked-env'
import { ChangeType, getChangedDocumentId, getActionType } from '../src/util'
import defaultEnvironment from './data/environment'

describe('getChangeType', () => {
  const firebaseMock = firebaseFunctionsTestInit()

  test('return a create change type', () => {
    const beforeSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      {},
      'docs/1'
    )
    const afterSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      { foo: 'bar' },
      'docs/1'
    )

    const documentChange = firebaseMock.makeChange(
      beforeSnapshot,
      afterSnapshot
    )

    const changeType: ChangeType = getActionType(documentChange)

    expect(changeType).toEqual(ChangeType.CREATE)
  })

  test('return a update change type', () => {
    const beforeSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      { foo: 'bar' },
      'docs/1'
    )
    const afterSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      { foo: 'bars' },
      'docs/1'
    )

    const documentChange = firebaseMock.makeChange(
      beforeSnapshot,
      afterSnapshot
    )

    const changeType: ChangeType = getActionType(documentChange)

    expect(changeType).toEqual(ChangeType.UPDATE)
  })

  test('return a delete change type', () => {
    const beforeSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      { foo: 'bar' },
      'docs/1'
    )
    const afterSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      {},
      'docs/1'
    )

    const documentChange = firebaseMock.makeChange(
      beforeSnapshot,
      afterSnapshot
    )

    const changeType: ChangeType = getActionType(documentChange)

    expect(changeType).toEqual(ChangeType.DELETE)
  })
})

describe('getChangedDocumentId', () => {
  const firebaseMock = firebaseFunctionsTestInit()

  test('return id after create document', () => {
    const beforeSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      {},
      'docs/1'
    )
    const afterSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      { foo: 'bar' },
      'docs/2'
    )

    const documentChange = firebaseMock.makeChange(
      beforeSnapshot,
      afterSnapshot
    )

    const id: string = getChangedDocumentId(documentChange)

    expect(id).toEqual('2')
  })

  test('return id after update document', () => {
    const beforeSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      { foo: 'bar' },
      'docs/1'
    )
    const afterSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      { foo: 'bars' },
      'docs/2'
    )

    const documentChange = firebaseMock.makeChange(
      beforeSnapshot,
      afterSnapshot
    )

    const id: string = getChangedDocumentId(documentChange)

    expect(id).toEqual('2')
  })

  test('return id after delete document', () => {
    const beforeSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      { foo: 'bar' },
      'docs/1'
    )
    const afterSnapshot = firebaseMock.firestore.makeDocumentSnapshot(
      {},
      'docs/2'
    )

    const documentChange = firebaseMock.makeChange(
      beforeSnapshot,
      afterSnapshot
    )

    const id: string = getChangedDocumentId(documentChange)

    expect(id).toEqual('1')
  })
})

describe('getFieldsToIndex', () => {
  let adapter
  let restoreEnv
  let mockParseFieldsToIndex
  const config = global.config

  beforeEach(() => {
    jest.resetModules()
    restoreEnv = mockedEnv(defaultEnvironment)
  })
  afterEach(() => restoreEnv())

  test('configuration detected from environment variables', () => {
    const mockConfig = config()
    expect(mockConfig).toMatchSnapshot()
  })

  test('return empty list', () => {
    adapter = require('../src/meilisearch-adapter')
    mockParseFieldsToIndex = adapter.parseFieldsToIndex()
    expect(mockParseFieldsToIndex).toMatchObject([])
  })

  test('return list with one field', () => {
    adapter = require('../src/meilisearch-adapter')
    mockParseFieldsToIndex = adapter.parseFieldsToIndex('field')
    expect(mockParseFieldsToIndex).toMatchObject(['field'])
  })

  test('return list with multiple fields', () => {
    adapter = require('../src/meilisearch-adapter')
    mockParseFieldsToIndex = adapter.parseFieldsToIndex('field1,field2,field3')
    expect(mockParseFieldsToIndex).toMatchObject(['field1', 'field2', 'field3'])
  })

  test('return list with multiple fields and spaces', () => {
    adapter = require('../src/meilisearch-adapter')
    mockParseFieldsToIndex = adapter.parseFieldsToIndex(
      'field1, field2,  field3'
    )
    expect(mockParseFieldsToIndex).toMatchObject(['field1', 'field2', 'field3'])
  })

  test('return list of fiels with underscore', () => {
    adapter = require('../src/meilisearch-adapter')
    mockParseFieldsToIndex = adapter.parseFieldsToIndex(
      'field_1,field_2,field_3'
    )
    expect(mockParseFieldsToIndex).toMatchObject([
      'field_1',
      'field_2',
      'field_3',
    ])
  })
})
