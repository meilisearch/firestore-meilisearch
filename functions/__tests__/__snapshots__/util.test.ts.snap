// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFieldsToIndex configuration detected from environment variables 1`] = `undefined`;
